File: Excel/800-899/887/887 Challenge.py
import pandas as pd
import numpy as np
import itertools
import re

path = "Excel/800-899/887/887 Minimum Product Triplet.xlsx"
input = pd.read_excel(path, usecols="A", skiprows=1, nrows=20)
test = pd.read_excel(path, usecols="B:C", skiprows=1, nrows=20)
test = test.applymap(lambda x: re.sub(r'\{', '(', str(x)))
test = test.applymap(lambda x: re.sub(r'\}', ')', str(x)))

def find_triplet(s):
    x = list(map(int, s.split(", ")))
    combs = list(itertools.combinations(x, 3))
    prods = [np.prod(c) for c in combs]
    m = min(prods)
    min_triplets = [tuple(sorted(combs[i])) for i, p in enumerate(prods) if p == m]
    unique_triplets = sorted(set(min_triplets), key=lambda t: sum(t))
    triplet_strs = [", ".join(map(str, t)) for t in unique_triplets]
    triplets_fmt = "(" + ") , (".join(triplet_strs) + ")"
    return {"Min Product": str(m), "Triplets": triplets_fmt}

result = pd.DataFrame([find_triplet(s) for s in input.iloc[:,0]])

print(result.equals(test))
# FALSE, two cells has wrong min product in solution. And there is some sorting problems

File: Excel/800-899/887/887 Challenge.R
library(tidyverse)
library(readxl)

path <- "Excel/800-899/887/887 Minimum Product Triplet.xlsx"
input <- read_excel(path, range = "A2:A22")
test <- read_excel(path, range = "B2:C22") %>%
  mutate_all(~ str_replace_all(., "\\{", "\\(")) %>%
  mutate_all(~ str_replace_all(., "\\}", "\\)"))


find_triplet = function(s) {
  x = as.numeric(strsplit(s, ", ")[[1]])
  c = combn(x, 3)
  m = min(p <- apply(c, 2, prod))
  t = unique(apply(apply(c[,p==m,drop=F],2,sort),2,paste,collapse=", "))
  t = t[order(rowSums(matrix(as.numeric(unlist(strsplit(t,", "))),ncol=3,byrow=T)))]
  list(`Min Product`=as.character(m), Triplets=paste0("(",paste(t,collapse=") , ("),")"))
}

result = map_dfr(input$Data, find_triplet)

all.equal(result, test, check.attributes = FALSE)
# FALSE, two cells has wrong min product in solution. And there is some sorting problems